src/main/kotlin/com/github/pedrovgs/gameoflife/model.kt

```diff
@@ -0,0 +1,115 @@
+package com.github.pedrovgs.gameoflife
+
+import java.lang.StringBuilder
+
+data class Cell(val isAlive: Boolean) {
+    companion object {
+        val alive: Cell = Cell(true)
+        val dead: Cell = Cell(false)
+    }
+
+    fun evolve(aliveNeighbours: Int): Cell =
+            if (isAlive) {
+                evolveLiveCell(aliveNeighbours)
+            } else {
+                evolveDeadCell(aliveNeighbours)
+            }
+
+    private fun evolveLiveCell(aliveNeighbours: Int): Cell =
+            when {
+                aliveNeighbours < 2 -> die()
+                aliveNeighbours == 2 || aliveNeighbours == 3 -> live()
+                else -> die()
+            }
+
+    private fun evolveDeadCell(aliveNeighbours: Int): Cell =
+            when (aliveNeighbours) {
+                3 -> live()
+                else -> die()
+            }
+
+    private fun live(): Cell = copy(isAlive = true)
+    private fun die(): Cell = copy(isAlive = false)
+}
+
+data class UniverseSize(val rows: Int, val columns: Int)
+data class Position(val x: Int, val y: Int)
+
+data class Universe(val cells: Array<Array<Cell>>) {
+    companion object {
+        fun create(size: UniverseSize, initialCells: Map<Position, Cell>) =
+                Universe((0 until size.rows).map { x ->
+                    (0 until size.columns).map { y ->
+                        initialCells[Position(x, y)] ?: Cell.dead
+                    }.toTypedArray()
+                }.toTypedArray())
+    }
+
+    val aliveCells = cells.map { it.count { cell -> cell.isAlive } }.sum()
+
+    val deadCells = cells.map { it.count() }.sum() - aliveCells
+
+    val numberOfCells = cells.map { it.count() }.sum()
+
+    fun tick(): Universe {
+        val newCellsAfterTick = cells.mapIndexed { x, rows ->
```

Comment:
Even if this is already merged: I'd say you can represent all tick transformations as S -> (S, A) which is State<S, A>. I think in the context of this problem A can be Unit, since you're only interested on the state itself.

Reply:
Thanks for the tip @JorgeCastilloPrz I thought that too 😃 However for my solution, I wanted to review the usage of the IO monad and keep the domain as simple as possible. I'll explore the usage of the State monad in a future PR. Basically, anything returning a new state can be represented with the State monad. However, I don't see any benefit if you are not going to mutate the state in the same block of code many times. I mean, for this exercise you could do it, but as this method is a pure function deferring the evaluation is something we do when wrapping the program into a bigger monad like IO when printing the state of the world. Maybe I'm missing any extra benefit 🤔

Reply:
I'm having lunch out of home so I can't look to the code in depth, but lemme throw some ideas here that might be useful for your explorations around State. Pls discard what you feel less relevant!

State also ensures (besides immutability) to automate the explicit state passing (instead of reading it from a val stored outside of the function scope). It makes it implicit, so you need mess the less possible with it, and to ensure the resulting state is used as the input of the next flatMap "iteration". Hence the program eforces you to model state change as sequential and linear (See http://eed3si9n.com/herding-cats/State.html) 

When I look at this problem I see an evolving program state that would be running infinitely (so you run it once then it's lazy evaluation keeps it alive forever, tick after tick without blowing up memory). That can be modelled with State to represent the tick() function (imposing all the domain rules under its calculation) then returning a new State with the result, then flatMap over a call to itself, so the resulting state of previous computation becomes the input of the next tick. State is effectively a function that requires an input state to calculate the result, so it's deferred by nature (waiting to have an input produced by previous tick).

Said that, you can indeed also model that in the same way but using IO and ensuring you don't mutate the state inside each tick with other techniques. The state approach is probably the most "academic" for a Kata to illustrate appliicagion state (which seems to fit fine for this one) and talso he first one that came to my mind when you think about how to model evolving app state, I guess. At the same time there's also the point that different Monads to not compose well, so if you really wanted to stay simple I guess running everything using IO is probably better.

A second alternative is exploring StateT so you can keep IO as your F but give it state handling capabilities. Not sure how that'd look tho.

Reply:
You always have a different point of view @JorgeCastilloPrz!! For this kata I think the usage of IO is enough, but let's do something interesting. For my next kata I'll try to use state monad and we will review the result. Ok? Thanks for your time and tips 😃 See you this Christmas in Murcia!

Reply:
Sounds great, see you soon 👌